src/snowflake/snowpark/_internal/compiler/utils.py

```diff
@@ -39,3 +58,86 @@ def create_query_generator(plan: SnowflakePlan) -> QueryGenerator:
         )
 
     return QueryGenerator(plan.session, snowflake_create_table_plan_info)
+
+
+def replace_child(
+    parent: PlanTreeNode, old_child: PlanTreeNode, new_child: SelectableEntity
+) -> None:
+    """
+    Helper function to replace the child node in the plan with a new selectable entity.
+    """
+    if not parent._is_valid_for_replacement:
+        raise ValueError(f"parent node {parent} is not valid for replacement.")
+
+    if old_child not in getattr(parent, "children_plan_nodes", parent.children):
+        raise ValueError(f"old_child {old_child} is not a child of parent {parent}.")
+
+    assert isinstance(
+        new_child, SelectableEntity
+    ), f"expecting new_child to be SelectableEntity, got {type(new_child)}"
+
+    if isinstance(parent, SnowflakePlan):
+        assert parent.source_plan is not None
+        replace_child(parent.source_plan, old_child, new_child)
+        return
+
+    if isinstance(parent, SelectStatement):
+        parent.from_ = new_child
+        return
+
+    if isinstance(parent, SetStatement):
+        parent._nodes = [
+            node if node != old_child else new_child for node in parent._nodes
+        ]
+        parent.set_operands = tuple(
+            operand
+            if operand.selectable != old_child
+            else SetOperand(new_child, operand.operator)
```

Comment:
instead of creating new one, we can direct update the selectable of the operand since it is an inplace update, in case other node is referring to the same operand